// @needle-tools/engine/lib/engine-components/postprocessing/Effects/Antialiasing.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { EdgeDetectionMode, SMAAEffect, SMAAPreset } from "postprocessing";
import { serializable } from "../../../engine/engine_serialization.js";
import { PostProcessingEffect } from "../PostProcessingEffect.js";
import { VolumeParameter } from "../VolumeParameter.js";
import { registerCustomEffectType } from "../VolumeProfile.js";
export var QualityLevel;
(function (QualityLevel) {
    QualityLevel[QualityLevel["LOW"] = 0] = "LOW";
    QualityLevel[QualityLevel["MEDIUM"] = 1] = "MEDIUM";
    QualityLevel[QualityLevel["HIGH"] = 2] = "HIGH";
    QualityLevel[QualityLevel["ULTRA"] = 3] = "ULTRA";
})(QualityLevel || (QualityLevel = {}));
/**
 * @category Effects
 */
export class Antialiasing extends PostProcessingEffect {
    get typeName() {
        return "Antialiasing";
    }
    // @serializable(VolumeParameter)
    // edgeDetectionThreshold!: VolumeParameter;
    preset = new VolumeParameter(SMAAPreset.HIGH);
    onCreateEffect() {
        const effect = new SMAAEffect({
            preset: SMAAPreset.HIGH,
            edgeDetectionMode: EdgeDetectionMode.DEPTH
        });
        this.preset.onValueChanged = (newValue) => {
            effect.applyPreset(newValue);
        };
        // effect.edgeDetectionMaterial.edgeDetectionThreshold = .01;
        // this.edgeDetectionThreshold.onValueChanged = (newValue) => {
        //     console.log(newValue);
        //     effect.edgeDetectionMaterial.edgeDetectionThreshold = newValue;
        // }
        return effect;
    }
}
__decorate([
    serializable(VolumeParameter)
], Antialiasing.prototype, "preset", void 0);
registerCustomEffectType("Antialiasing", Antialiasing);
