// @needle-tools/engine/src/engine-components/EventTrigger.ts
import { serializable } from "../engine/engine_serialization.js";
import { Behaviour } from "./Component.js"
import { EventList } from "./EventList.js";
import { EventType } from "./EventType.js"
import type { IPointerEventHandler, PointerEventData } from "./ui/PointerEvents.js"

class TriggerEvent {
    @serializable()
    eventID!: EventType;
    @serializable(EventList)
    callback?: EventList;
}

/**
 * The EventTrigger component is used to trigger events when certain pointer events occur on the GameObject.
 * It implements the {@link IPointerEventHandler} interface and can be used to expose events to the user in the editor without writing code.
 * @category Interactivity
 */
export class EventTrigger extends Behaviour implements IPointerEventHandler {

    /** A list of events that should be triggered when a pointer event occurs on the GameObject. */
    @serializable(TriggerEvent)
    triggers?: Array<TriggerEvent> = [];

    /** @internal */
    invoke(type: EventType) {
        if (!this.triggers) return;
        for (const trigger of this.triggers) {
            if (trigger.eventID === type) {
                trigger.callback?.invoke();
            }
        }
    }

    /** @internal */
    onPointerClick(_: PointerEventData) {
        this.invoke(EventType.PointerClick);
    }

    /** @internal */
    onPointerEnter(_: PointerEventData) {
        this.invoke(EventType.PointerEnter);
    }

    /** @internal */
    onPointerExit(_: PointerEventData) {
        this.invoke(EventType.PointerExit);
    }

    /** @internal */
    onPointerDown(_: PointerEventData) {
        this.invoke(EventType.PointerDown);
    }

    /** @internal */
    onPointerUp(_: PointerEventData) {
        this.invoke(EventType.PointerUp);
    }

}
