// @needle-tools/engine/plugins/vite/buildinfo.js
import { createBuildInfoFile } from '../common/buildinfo.js';
import { buildPipelineTask } from './build-pipeline.js';
import { getOutputDirectory } from './config.js';



/** Create a buildinfo file in the build directory
 * @param {import('../types').userSettings} userSettings
 * @returns {import('vite').Plugin}
 */
export const needleBuildInfo = (command, config, userSettings) => {

    if (userSettings?.noBuildInfo) return;

    return {
        name: 'needle:buildinfo',
        apply: "build",
        enforce: "post",
        closeBundle: async () => {
            return new Promise(async res => {
                if (buildPipelineTask instanceof Promise) {
                    await buildPipelineTask.catch(() => { });
                }
                // wait for gzip
                await delay(500);
                const buildDirectory = getOutputDirectory();
                createBuildInfoFile(buildDirectory);
                res();
            });
        }
    }
}

function delay(ms) {
    return new Promise(res => setTimeout(res, ms));
}
