// @needle-tools/engine/src/engine-schemes/dist/vec4.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
exports.__esModule = true;
exports.Vec4 = void 0;
var Vec4 = /** @class */ (function () {
    function Vec4() {
        this.bb = null;
        this.bb_pos = 0;
    }
    Vec4.prototype.__init = function (i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    };
    Vec4.prototype.x = function () {
        return this.bb.readFloat32(this.bb_pos);
    };
    Vec4.prototype.y = function () {
        return this.bb.readFloat32(this.bb_pos + 4);
    };
    Vec4.prototype.z = function () {
        return this.bb.readFloat32(this.bb_pos + 8);
    };
    Vec4.prototype.w = function () {
        return this.bb.readFloat32(this.bb_pos + 12);
    };
    Vec4.sizeOf = function () {
        return 16;
    };
    Vec4.createVec4 = function (builder, x, y, z, w) {
        builder.prep(4, 16);
        builder.writeFloat32(w);
        builder.writeFloat32(z);
        builder.writeFloat32(y);
        builder.writeFloat32(x);
        return builder.offset();
    };
    return Vec4;
}());
exports.Vec4 = Vec4;
